File: CapitalesAfricaJava/app/src/main/java/com/example/capitalesafricajava/Juego.java
package com.example.capitalesafricajava;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.snackbar.Snackbar;

import java.util.ArrayList;
import java.util.LinkedHashMap;

public class Juego extends AppCompatActivity {

    LinkedHashMap<String, String> capitalesPaises = new LinkedHashMap<>();
    ArrayList<String> paises;
    ArrayList<String> capitales;
    ArrayList<String> paisesNoJugados = new ArrayList<>();
    int puntuacion = 0;
    int cantidad = 0;
    String respuesta;
    DbHandler dbHandler = new DbHandler(Juego.this);
    boolean control = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.juego);
        llenaMapa();
        ponePais();
        ImageButton jugar = (ImageButton) findViewById(R.id.btnJugar);
        RadioGroup grupo = (RadioGroup) findViewById(R.id.radioGroup);
        TextView debug = (TextView) findViewById(R.id.textView3);
        TextView pais = (TextView) findViewById(R.id.txtPais);

        jugar.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                int checkedRadioButtonId = grupo.getCheckedRadioButtonId();
                RadioButton seleccionado = findViewById(checkedRadioButtonId);
                if (checkedRadioButtonId == -1) {
                    // No item selected
                }
//                else if (cantidad == 9) {

//                }
                else {
                    cantidad++;
                    if (seleccionado.getText().equals(respuesta)) {
                        puntuacion++;
                        dbHandler.insertUserDetails((String) pais.getText(), respuesta, "bien");
                        try {
                            sacaToast("Correcto!!", "GREEN");
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        if (!(cantidad > 9)) {
                            ponePais();
                        } else {
                            Intent resultado = new Intent(Juego.this, Resultado.class);
                            startActivity(resultado);
                        }

                    } else {
                        dbHandler.insertUserDetails((String) pais.getText(), respuesta, "mla");
                        try {
                            sacaToast("Error", "RED");
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        if (!(cantidad > 9)) {
                            ponePais();
                        } else {
                            Intent resultado = new Intent(Juego.this, Resultado.class);
                            startActivity(resultado);
                        }
                    }
                    grupo.clearCheck();


                }
            }

        });
    }


    public void llenaMapa() {
        capitalesPaises.put("Marruecos", "Rabat");
        capitalesPaises.put("Egipto", "El Cairo");
        capitalesPaises.put("Mali", "Bamako");
        capitalesPaises.put("Niger", "Niamey");
        capitalesPaises.put("Angola", "Luanda");
        capitalesPaises.put("Bostwana", "Gaborone");
        capitalesPaises.put("Namibia", "Windhoek");
        capitalesPaises.put("Etiopía", "Addis Ababa");
        capitalesPaises.put("Nigeria", "Abuja");
        capitalesPaises.put("Chad", "N'Djamena");
        paises = new ArrayList<String>(capitalesPaises.keySet());
        paisesNoJugados = new ArrayList<String>(capitalesPaises.keySet());
        capitales = new ArrayList<String>(capitalesPaises.values());
    }

    public void ponePais() {

        String p = "";
        String radio1;
        String radio2;
        String radio3;
        String radio4;
        TextView pais = (TextView) findViewById(R.id.txtPais);
        RadioButton r1 = (RadioButton) findViewById(R.id.radio1);
        RadioButton r2 = (RadioButton) findViewById(R.id.radio2);
        RadioButton r3 = (RadioButton) findViewById(R.id.radio3);
        RadioButton r4 = (RadioButton) findViewById(R.id.radio4);
        int aleat = (int) (Math.random() * paisesNoJugados.size());
        p = paisesNoJugados.get(aleat);
        paisesNoJugados.remove(aleat);
        respuesta = capitalesPaises.get(p);
        pais.setText(p);
        radio1 = sacaCapital();
        r1.setText(radio1);
        do {
            radio2 = sacaCapital();
        } while (radio2.equals(radio1));
        r2.setText(radio2);
        do {
            radio3 = sacaCapital();
        } while (radio2.equals(radio3) | radio1.equals(radio3));

        r3.setText(radio3);

        do {
            if (!radio2.equals(capitalesPaises.get(p)) && !radio1.equals(capitalesPaises.get(p)) && !radio3.equals(capitalesPaises.get(p))) {
                radio4 = capitalesPaises.get(p);
            } else {
                radio4 = sacaCapital();
            }
        } while ((radio2.equals(radio4) | radio1.equals(radio4) | radio3.equals(radio4)));
        r4.setText(radio4);
    }

    public String sacaCapital() {
        int aleat = (int) (Math.random() * 9);
        return capitales.get(aleat);
    }

    // Función que saca los toast
    private void sacaToast(String texto, String color) throws InterruptedException {
        Toast toast = Toast.makeText(
                getApplicationContext(),
                texto,
                Toast.LENGTH_SHORT
        );
        toast.setGravity(Gravity.CENTER_VERTICAL | Gravity.CENTER_VERTICAL, 0, 0);
        toast.getView().setBackgroundColor(Color.parseColor(color));
        toast.show();

    }


}

File: CapitalesAfricaKotlin/app/src/main/java/com/example/capitalesafricakotlin/Juego.kt
package com.example.capitalesafricajava

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import java.util.*

class Juego : AppCompatActivity() {
    var capitalesPaises = LinkedHashMap<String, String>()
    var paises: ArrayList<String>? = null
    var capitales: ArrayList<String>? = null
    var paisesNoJugados = ArrayList<String>()
    var puntuacion = 0
    var cantidad = 0
    var respuesta: String? = null
    var dbHandler = DbHandler(this@Juego)
    var control = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.juego)
        llenaMapa()
        ponePais()
        val jugar = findViewById<View>(R.id.btnJugar) as ImageButton
        val grupo = findViewById<View>(R.id.radioGroup) as RadioGroup
        val debug = findViewById<View>(R.id.textView3) as TextView
        val pais = findViewById<View>(R.id.txtPais) as TextView
        jugar.setOnClickListener {
            val checkedRadioButtonId = grupo.checkedRadioButtonId
            val seleccionado = findViewById<RadioButton>(checkedRadioButtonId)
            if (checkedRadioButtonId == -1) {
                // No item selected
            } else {
                cantidad++
                if (seleccionado.text == respuesta) {
                    puntuacion++
                    dbHandler.insertUserDetails(pais.text as String, respuesta, "bien")
                    try {
//                        sacaToast("Correcto!!", "GREEN")
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                    if (cantidad <= 9) {
                        ponePais()
                    } else {
                        val resultado = Intent(this@Juego, Resultado::class.java)
                        startActivity(resultado)
                    }
                } else {
                    dbHandler.insertUserDetails(pais.text as String, respuesta, "mla")
                    try {
//                        sacaToast("Error", "RED")
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                    if (cantidad <= 9) {
                        ponePais()
                    } else {
                        val resultado = Intent(this@Juego, Resultado::class.java)
                        startActivity(resultado)
                    }
                }
                grupo.clearCheck()
            }
        }
    }

    fun llenaMapa() {
        capitalesPaises["Marruecos"] = "Rabat"
        capitalesPaises["Egipto"] = "El Cairo"
        capitalesPaises["Mali"] = "Bamako"
        capitalesPaises["Niger"] = "Niamey"
        capitalesPaises["Angola"] = "Luanda"
        capitalesPaises["Bostwana"] = "Gaborone"
        capitalesPaises["Namibia"] = "Windhoek"
        capitalesPaises["Etiopía"] = "Addis Ababa"
        capitalesPaises["Nigeria"] = "Abuja"
        capitalesPaises["Chad"] = "N'Djamena"
        paises = ArrayList(capitalesPaises.keys)
        paisesNoJugados = ArrayList(capitalesPaises.keys)
        capitales = ArrayList(capitalesPaises.values)
    }

    fun ponePais() {
        var p = ""
        val radio1: String
        var radio2: String
        var radio3: String
        var radio4: String?
        val pais = findViewById<View>(R.id.txtPais) as TextView
        val r1 = findViewById<View>(R.id.radio1) as RadioButton
        val r2 = findViewById<View>(R.id.radio2) as RadioButton
        val r3 = findViewById<View>(R.id.radio3) as RadioButton
        val r4 = findViewById<View>(R.id.radio4) as RadioButton
        val aleat = (Math.random() * paisesNoJugados.size).toInt()
        p = paisesNoJugados[aleat]
        paisesNoJugados.removeAt(aleat)
        respuesta = capitalesPaises[p]
        pais.text = p
        radio1 = sacaCapital()
        r1.text = radio1
        do {
            radio2 = sacaCapital()
        } while (radio2 == radio1)
        r2.text = radio2
        do {
            radio3 = sacaCapital()
        } while ((radio2 == radio3) or (radio1 == radio3))
        r3.text = radio3
        do {
            radio4 =
                if (radio2 != capitalesPaises[p] && radio1 != capitalesPaises[p] && radio3 != capitalesPaises[p]
                ) {
                    capitalesPaises[p]
                } else {
                    sacaCapital()
                }
        } while ((radio2 == radio4) or (radio1 == radio4) or (radio3 == radio4))
        r4.text = radio4
    }

    fun sacaCapital(): String {
        val aleat = (Math.random() * 9).toInt()
        return capitales!![aleat]
    }

    // Función que saca los toast
    @Throws(InterruptedException::class)
    private fun sacaToast(texto: String, color: String) {
        val toast = Toast.makeText(
            applicationContext,
            texto,
            Toast.LENGTH_SHORT
        )
        toast.setGravity(Gravity.CENTER_VERTICAL or Gravity.CENTER_VERTICAL, 0, 0)
        toast.view!!.setBackgroundColor(Color.parseColor(color))
        toast.show()
    }
}